### PR TITLE
Explain k in approximate k-NN

### LINKED ISSUE #7108: [DOC] number of results for k is unclear
[The docs say:](https://opensearch.org/docs/latest/query-dsl/specialized/neural/)

k 	Integer 	Optional 	The number of results returned by the k-NN search. Default is 10.

It appears that k is actually the number of results *per shard* not total, though I'm not totally sure.  This seemed unclear from earlier

### PR DESCRIPTION
**Pull request by kolchfa-aws**
Closes #7108 

### Checklist
- [x] By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license and subject to the [Developers Certificate of Origin](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: _search-plugins/knn/approximate-knn.md [hunk 1]
```diff
@@ -127,10 +127,14 @@ GET my-knn-index-1/_search
 }
 ```
 
-`k` is the number of neighbors the search of each graph will return. You must also include the `size` option, which
-indicates how many results the query actually returns. The plugin returns `k` amount of results for each shard
-(and each segment) and `size` amount of results for the entire query. The plugin supports a maximum `k` value of 10,000.
-Starting in OpenSearch 2.14, in addition to using the `k` variable, both the `min_score` and `max_distance` variables can be used for [radial search]({{site.url}}{{site.baseurl}}/search-plugins/knn/radial-search-knn/).
+In the preceding query, `k` is the number of neighbors the search of each graph returns. You must also include the `size` option, indicating the final number of results you want the query to return.  
+
+For NMSLIB and Faiss engines, `k` is the maximum number of documents returned for all segments of a shard. For the Lucene engine, `k` is the number of documents returned for a shard. The maximum value of `k` is 10,000.
```
Comment (navneet1v) [inline _search-plugins/knn/approximate-knn.md:132]: This line is not adding no new information if we don't add an example. An example like this:

| size | k  | Number of Shards | Number of segments per shard | Engine       | Final Number of Results Returned |
|------|----|------------------|------------------------------|--------------|----------------------------------|
| 10   | 1  | 1                | 4                            | Faiss/Nmslib | 4                                |
| 10   | 1  | 1                | 4                            | Lucene       | 1                                |
| 10   | 10 | 1                | 4                            | Faiss/Nmslib | 10                               |
| 10   | 10 | 1                | 4                            | Lucene       | 10                               |
| 10   | 1  | 2                | 4                            | Faiss/Nmslib | 8                                |
| 10   | 1  | 2                | 4                            | Lucene       | 2                                |

Comment (kolchfa-aws) [reply _search-plugins/knn/approximate-knn.md:134]: @navneet1v Thank you for the additional information. I have added it to the docs. Please review again.

File: _search-plugins/knn/approximate-knn.md [hunk 2]
```diff
@@ -127,10 +127,25 @@ GET my-knn-index-1/_search
...
-Starting in OpenSearch 2.14, in addition to using the `k` variable, both the `min_score` and `max_distance` variables can be used for [radial search]({{site.url}}{{site.baseurl}}/search-plugins/knn/radial-search-knn/).
+### The number of returned results
+
+In the preceding query, `k` is the number of neighbors the search of each graph returns. You must also include the `size` option, indicating the final number of results you want the query to return.  
+
+For NMSLIB and Faiss engines, `k` is the maximum number of documents returned for all segments of a shard. For the Lucene engine, `k` is the number of documents returned for a shard. The maximum value of `k` is 10,000.
+
+For any engine, each shard returns `size` results to the coordinator node. Thus, the total number of results the coordinator node receives is `size * number of shards`. After the coordinator node consolidates the results received from all nodes, the query returns the top `size` results.
+
+The following table provides examples of the number of results returned by various engines in several scenarios.
```
Comment (navneet1v) [inline _search-plugins/knn/approximate-knn.md:138]: @kolchfa-aws why we are not adding the full table here?

Comment (kolchfa-aws) [reply _search-plugins/knn/approximate-knn.md:138]: @navneet1v It is the full table, but each pair of two identical rows is one row that contains Faiss/NMSLIB and Lucene results side by side so it's easier for the users to compare those.

Comment (navneet1v) [reply _search-plugins/knn/approximate-knn.md:138]: oh.. make sense. So you added 1 more column. thanks

Comment (navneet1v) [reply _search-plugins/knn/approximate-knn.md:138]: Can we add 1 more line above this, saying that this the table assumes data there are enough number of documents across shards and segments.

Comment (kolchfa-aws) [reply _search-plugins/knn/approximate-knn.md]: @navneet1v Done!

**Review by navneet1v — APPROVED** [review]

**Comment by navneet1v** [comment]
@kolchfa-aws approved from my side.

File: _search-plugins/knn/approximate-knn.md [hunk 3]
```diff
@@ -127,10 +127,25 @@ GET my-knn-index-1/_search
 }
 ```
 
-`k` is the number of neighbors the search of each graph will return. You must also include the `size` option, which
-indicates how many results the query actually returns. The plugin returns `k` amount of results for each shard
-(and each segment) and `size` amount of results for the entire query. The plugin supports a maximum `k` value of 10,000.
-Starting in OpenSearch 2.14, in addition to using the `k` variable, both the `min_score` and `max_distance` variables can be used for [radial search]({{site.url}}{{site.baseurl}}/search-plugins/knn/radial-search-knn/).
+### The number of returned results
+
+In the preceding query, `k` is the number of neighbors the search of each graph returns. You must also include the `size` option, indicating the final number of results you want the query to return.
```
Comment (natebower) [inline _search-plugins/knn/approximate-knn.md:132]:
```suggestion
In the preceding query, `k` represents the number of neighbors returned by the search of each graph. You must also include the `size` option, indicating the final number of results that you want the query to return.  
```

File: _search-plugins/knn/approximate-knn.md [hunk 4]
```diff
@@ -127,10 +127,25 @@ GET my-knn-index-1/_search
...
 
-`k` is the number of neighbors the search of each graph will return. You must also include the `size` option, which
-indicates how many results the query actually returns. The plugin returns `k` amount of results for each shard
-(and each segment) and `size` amount of results for the entire query. The plugin supports a maximum `k` value of 10,000.
-Starting in OpenSearch 2.14, in addition to using the `k` variable, both the `min_score` and `max_distance` variables can be used for [radial search]({{site.url}}{{site.baseurl}}/search-plugins/knn/radial-search-knn/).
+### The number of returned results
+
+In the preceding query, `k` is the number of neighbors the search of each graph returns. You must also include the `size` option, indicating the final number of results you want the query to return.  
+
+For NMSLIB and Faiss engines, `k` is the maximum number of documents returned for all segments of a shard. For the Lucene engine, `k` is the number of documents returned for a shard. The maximum value of `k` is 10,000.
```
Comment (natebower) [inline _search-plugins/knn/approximate-knn.md:134]:
```suggestion
For NMSLIB and Faiss engines, `k` represents the maximum number of documents returned for all segments of a shard. For the Lucene engine, `k` represents the number of documents returned for a shard. The maximum value of `k` is 10,000.
```

Comment (natebower) [inline _search-plugins/knn/approximate-knn.md:134]: Should "the" precede "NMSLIB and Faiss", as it does "Lucene"?

File: _search-plugins/knn/approximate-knn.md [hunk 5]
```diff
@@ -127,10 +127,25 @@ GET my-knn-index-1/_search
...
+
+For any engine, each shard returns `size` results to the coordinator node. Thus, the total number of results the coordinator node receives is `size * number of shards`. After the coordinator node consolidates the results received from all nodes, the query returns the top `size` results.
```
Comment (natebower) [inline _search-plugins/knn/approximate-knn.md:136]:
```suggestion
For any engine, each shard returns `size` results to the coordinator node. Thus, the total number of results that the coordinator node receives is `size * number of shards`. After the coordinator node consolidates the results received from all nodes, the query returns the top `size` results.
```

File: _search-plugins/knn/approximate-knn.md [hunk 6]
```diff
@@ -127,10 +127,25 @@ GET my-knn-index-1/_search
...
+
+The following table provides examples of the number of results returned by various engines in several scenarios. For these examples, assume that the number of documents in the segments and shards is sufficient to return the number of results specified in the table.
```
Comment (natebower) [inline _search-plugins/knn/approximate-knn.md:138]:
```suggestion
The following table provides examples of the number of results returned by various engines in several scenarios. For these examples, assume that the number of documents contained in the segments and shards is sufficient to return the number of results specified in the table.
```

File: _search-plugins/knn/approximate-knn.md [hunk 7]
```diff
@@ -127,10 +127,25 @@ GET my-knn-index-1/_search
...
+
+`size` 	| `k` | Number of primary shards | 	Number of segments per shard | Number of returned results, Faiss/NMSLIB | Number of returned results, Lucene
```
Comment (natebower) [inline _search-plugins/knn/approximate-knn.md:140]: "Number of returned Faiss/NMSLIB results"? "Number of returned Lucene results"?

File: _search-plugins/knn/approximate-knn.md [hunk 8]
```diff
@@ -127,10 +127,25 @@ GET my-knn-index-1/_search
...
+10 |	1 |	1 |	4 |	4 | 1
+10 | 10 |	1 |	4 |	10 | 10
+10 |	1 |	2 |	4 |	8 | 2
+ 
+Faiss/NMSLIB return the number of results different from Lucene only when `k` is smaller than `size`. If `k` and `size` are equal, all engines return the same number of results.
```
Comment (natebower) [inline _search-plugins/knn/approximate-knn.md:145]:
```suggestion
Faiss/NMSLIB return the number of results differently from Lucene only when `k` is smaller than `size`. If `k` and `size` are equal, all engines return the same number of results. 
```

File: _search-plugins/knn/approximate-knn.md [hunk 9]
```diff
@@ -127,10 +127,25 @@ GET my-knn-index-1/_search
...
+The following table provides examples of the number of results returned by various engines in several scenarios. For these examples, assume that the number of documents in the segments and shards is sufficient to return the number of results specified in the table.
+
+`size` 	| `k` | Number of primary shards | 	Number of segments per shard | Number of returned results, Faiss/NMSLIB | Number of returned results, Lucene
+10 |	1 |	1 |	4 |	4 | 1
+10 | 10 |	1 |	4 |	10 | 10
+10 |	1 |	2 |	4 |	8 | 2
+ 
+Faiss/NMSLIB return the number of results different from Lucene only when `k` is smaller than `size`. If `k` and `size` are equal, all engines return the same number of results. 
+
+Starting with OpenSearch 2.14, you can use `k`, `min_score`, or `max_distance` for [radial search]({{site.url}}{{site.baseurl}}/search-plugins/knn/radial-search-knn/).
```
Comment (natebower) [inline _search-plugins/knn/approximate-knn.md:147]:
```suggestion
Starting in OpenSearch 2.14, you can use `k`, `min_score`, or `max_distance` for [radial search]({{site.url}}{{site.baseurl}}/search-plugins/knn/radial-search-knn/).
```

**Review by natebower — COMMENTED** [review]
@kolchfa-aws Please see my comments and changes and let me know if you have any questions. Thanks!

File: _search-plugins/knn/approximate-knn.md (hunk 7 above)
Comment (kolchfa-aws) [reply _search-plugins/knn/approximate-knn.md:140]: I think it's clearer if you put the engine last.

**Review by Naarcha-AWS — APPROVED** [review]

**Pull request merged**